uxarray/grid.py

```diff
@@ -0,0 +1,279 @@
+        except ValueError as e:
+            # initialize from vertices
+            if in_type is np.ndarray:
+                self.vertices = args[0]
+                self.from_vert()
+            else:
+                print(
+                    "Init called with args other than verts (ndarray) or filename (str)"
+                )
+                exit()
+
+        # check if initialize from file:
+        if in_type is str and os.path.isfile(args[0]):
+            self.filepath = args[0]
+            self.from_file()
+
+        # initialize for gridspec
+        elif in_type is str:
+            self.gridspec = args[0]
+            self.from_gridspec()
+
+        else:
+            # this may just be local initialization for with no options or options other than above
+            pass
+
+    # vertices init
+    def from_vert(self):
+        """Create a grid with one face with vertices specified by the given
+        argument."""
+        self._init_mesh2()
+        self.in_ds.Mesh2.attrs['topology_dimension'] = self.vertices[0].size
+
+        self.in_ds["Mesh2"].topology_dimension
+
+        x_coord = self.vertices.transpose()[0]
+        y_coord = self.vertices.transpose()[1]
+
+        # single face with all nodes
+        num_nodes = x_coord.size
+        conn = list(range(0, num_nodes))
+        conn = [conn]
+
+        self.in_ds["Mesh2_node_x"] = xr.DataArray(data=xr.DataArray(x_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_node_y"] = xr.DataArray(data=xr.DataArray(y_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_face_nodes"] = xr.DataArray(
+            data=xr.DataArray(conn),
+            dims=["nMesh2_face", "nMaxMesh2_face_nodes"],
+            attrs={
+                "cf_role": "face_node_connectivity",
+                "_FillValue": -1,
+                "start_index": 0
+            })
+
+    # TODO: gridspec init
+    def from_gridspec(self):
+        print("initializing with gridspec")
+
+    # load mesh from a file
+    def from_file(self):
+        """Loads a mesh file Also, called by __init__ routine This routine will
+        automatically detect if it is a UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+        # find the file type
+        self.meshFileType = self.find_type()
+
+        # call reader as per meshFileType
+        if self.meshFileType == "exo":
+            self.read_exodus(self.ext_ds)
+        elif self.meshFileType == "scrip":
+            self.read_scrip(self.ext_ds)
+        elif self.meshFileType == "ugrid":
+            self.read_ugrid(self.filepath)
+        elif self.meshFileType == "shp":
+            self.read_shpfile(self.filepath)
+        else:
+            raise RuntimeError("unknown file format:" + self.meshFileType)
+
+    # helper function to find file type
+    def find_type(self):
```

Comment:
I believe this helper function is intended for internal use only. If that is correct, please name it with a leading single underscore as [suggested by PEP8](https://pep8.org/#:~:text=internal%20interfaces%20(packages%2C%20modules%2C%20classes%2C%20functions%2C%20attributes%20or%20other%20names)%20should%20still%20be%20prefixed%20with%20a%20single%20leading%20underscore). This is also applicable to any other internal API function in this PR, if any.

Reply:
Thanks for the commit to address this, but I think the above linked PEP8 doc suggests that the internal API functions use this convention instead: `self._find_type()` (i.e. prefixed with a single leading underscore). I might be mistaken though.